### ISSUE TITLE
2021-10-30

### ISSUE THREAD

**Issue by ssfc** [ssfc/wangdao#640]
211一个核心的词汇叫做：Contrastive。这个词翻译成中文是 对比 的意思，它的实质就是：试图教机器区分相似和不相似的事物。/// 比如说现在我们有任意的 4 张 images，如下图4所示。前两张都是dog 这个类别，后两张是其他类别，以第1张图为例，我们就希望它与第2张图的相似性越高越好，而与第3，第4张图的相似性越低越好。
212我们只有大堆images，没有任何标签，不知道哪些是 dog 这个类的，哪些是其他类的。没办法找出哪些图片应该去 Maximize Similarity，哪些应该去 Minimize Similarity。
213假设现在有1张任意的图片 [公式] ，叫做Original Image，先对它做数据增强，得到2张增强以后的图片 [公式] 。注意数据增强的方式有以下3种：随机裁剪之后再resize成原来的大小 (Random cropping followed by resize back to the original size)。随机色彩失真 (Random color distortions)。随机高斯模糊 (Random Gaussian Deblur)。接下来把增强后的图片 [公式] 输入到Encoder里面，注意这2个Encoder是共享参数的，得到representation [公式] ，再把 [公式] 继续通过 Projection head 得到 representation [公式] ，这里的2个 Projection head 依旧是共享参数的，且其具体的结构表达式是：[公式] 接下来的目标就是最大化同一张图片得到的 [公式] 。
2141.1 数据增强; 随机裁剪之后再resize成原来的大小 (Random cropping followed by resize back to the original size); 随机色彩失真 (Random color distortions); 随机高斯模糊 (Random Gaussian Deblur); 
215使用预测头 Projection head。在 SimCLR 中，Encoder 得到的2个 visual representation再通过Prediction head ([公式])进一步提特征，预测头是一个 2 层的MLP，将 visual representation 这个 2048 维的向量[公式]进一步映射到 128 维隐空间中，得到新的representation [公式]。利用 [公式] 去求loss 完成训练，训练完毕后扔掉预测头，保留 Encoder 用于获取 visual representation。
216首先定义Representation之间的相似度：使用余弦相似度Cosine Similarity; 使用上述公式计算batch里面的每个Augmented Images [公式] 的成对余弦相似度。 如下图13所示，在理想情况下，狗的增强图像之间的相似度会很高，而狗和鲸鱼图像之间的相似度会较低。
217数据增强是数据扩大的意思; (2021-10-20)
218SimCLR用了一种叫做 NT-Xent loss (Normalized Temperature-Scaled Cross-Entropy Loss)的对比学习损失函数。
219这种 softmax 计算等价于获得第2张增强的狗的图像与该对中的第1张狗的图像最相似的概率。 在这里，分母中的其余的项都是其他图片的增强之后的图片，也是negative samples。// 所以我们希望上面的softmax的结果尽量大，所以损失函数取了softmax的负对数： (2021-10-20)
